### PR TITLE
EdgeHub: Twin manager 

### PR DESCRIPTION
**Pull request by varunpuranik**
Adding a storing twin manager and pass through twin manager. 

**Pending:**
- Working on tests - will add in next PR
- Case when a client sends reported properties when offline and then goes away.

### CODE REVIEW

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/AssemblyInfo.cs [hunk 1]
```diff
@@ -4,3 +4,4 @@
 [assembly: InternalsVisibleTo("Microsoft.Azure.Devices.Edge.Hub.Core.Test")]
 [assembly: InternalsVisibleTo("Microsoft.Azure.Devices.Edge.Hub.Mqtt.Test")]
 [assembly: InternalsVisibleTo("Microsoft.Azure.Devices.Edge.Hub.E2E.Test")]
+[assembly: InternalsVisibleTo("DynamicProxyGenAssembly2")]
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/AssemblyInfo.cs]: What's this for?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+            }
+
+            this.syncToCloudSignal.Set();
+        }
+
+        public async Task SyncToCloud(string id)
+        {
+            Events.SyncingReportedPropertiesToCloud(id);
+            Option<TwinStoreEntity> twinWithReportedProperties = (await this.twinStore.Get(id))
+                .Filter(t => t.ReportedPropertiesPatch.HasValue);
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs]: The twin store contains the latest patch?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                            await ti.ReportedPropertiesPatch.ForEachAsync(
+                                async reportedPropertiesPatch =>
+                                {
+                                    bool result = await this.cloudSync.UpdateReportedProperties(id, reportedPropertiesPatch);
+                                    if (result)
+                                    {
+                                        Events.UpdateReportedPropertiesSucceeded(id);
+                                        await this.twinStore.Update(
+                                            id,
+                                            t => new TwinStoreEntity(t.Twin, Option.None<TwinCollection>()));
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs]: Then, once synced to the cloud, it is set to `None`?

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                }
+                catch (Exception e)
+                {
+                    Events.ErrorSyncingReportedPropertiesToCloud(e);
+                }
+
+                // Wait for syncfrequency to avoid looping too fast,
+                // then wait for the signal indicating more work is ready
+                await Task.Delay(this.syncFrequency);
+                await this.syncToCloudSignal.WaitAsync();
```
Comment (myagley) [inline edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs]: Who sets this signal? I see it in the InitSync call. Anywhere else?

Comment (myagley) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs]: Ah. It's set externally.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/AssemblyInfo.cs (hunk 1 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/AssemblyInfo.cs]: To allow mocking Internal interfaces with Moq

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs (hunk 2 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs]: It contains all the unapplied patches (which are applied on top of one another). 
Say if the device sends the following patches in quick succession (before any sync to IoTHub happens) - rp1, rp2, rp3 - then the store will contain the patches applied on top of one another - rp_store = { rp1 + rp2 + rp3 }

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs (hunk 3 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs]: Right - the stored entity contains the reported properties that need to be synced to the cloud. Once that is done, the entry is cleared out. 
The entity looks like this - 
{
   Twin { ReportedProperties, DesiredProperties },
   ReportedPropertiesToSync
}
So, in this operation, the reportedPropertiesToSync is set to None.

File: edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs (hunk 4 above)
Comment (varunpuranik) [reply edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/twin/ReportedPropertiesStore.cs]: Right, every time a client updates its reported properties, we signal this loop to get going again.

**Comment by varunpuranik** [comment]
> How are we handling backwards compatibility? Did the schema of the entities in the store change?
The schema did change, but I have backward compat tests :)
https://github.com/Azure/iotedge/pull/660/files#diff-6a1689f378bb49e2a7e9d472c714915bR61

**Pull request merged**